compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1567,6 +1573,92 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    /// Given a type mismatch error caused by `&T` being cloned instead of `T`, and
+    /// the `expr` as the source of this type mismatch, try to find the method call
+    /// as the source of this error and return that instead. Otherwise, return the
+    /// original expression.
+    fn note_type_is_not_clone_inner_expr<'b>(
+        &'b self,
+        expr: &'b hir::Expr<'b>,
+    ) -> &'b hir::Expr<'b> {
+        match expr.kind {
+            hir::ExprKind::Path(hir::QPath::Resolved(
+                None,
+                hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
+            )) => {
+                let Some(hir::Node::Pat(hir::Pat { hir_id, .. })) = self.tcx.hir().find(*binding)
+                else {
+                    return expr;
+                };
+                let Some(parent) = self.tcx.hir().find(self.tcx.hir().parent_id(*hir_id)) else {
+                    return expr;
+                };
+
+                match parent {
+                    // foo.clone()
+                    hir::Node::Local(hir::Local { init: Some(init), .. }) => {
+                        self.note_type_is_not_clone_inner_expr(init)
+                    }
+                    // When `expr` is more complex like a tuple
+                    hir::Node::Pat(hir::Pat { hir_id: pat_hir_id, kind: hir::PatKind::Tuple(pats, ..), .. }) => {
+                        let Some(hir::Node::Local(hir::Local { init: Some(init), .. })) =
+                                self.tcx.hir().find(self.tcx.hir().parent_id(*pat_hir_id)) else {
+                            return expr;
+                        };
+
+                        let find_init = |init_tup: &'b [Expr<'b>]| {
+                            pats.iter().enumerate().filter(|x| x.1.hir_id == *hir_id).map(|(i,_)| init_tup.get(i).unwrap()).next()
+                        };
+
+                        match init.kind {
+                            ExprKind::Tup(init_tup) => {
+                                if let Some(init) = find_init(init_tup) {
+                                    self.note_type_is_not_clone_inner_expr(init)
+                                } else {
+                                    expr
+                                }
+                            }
+                            ExprKind::Block(_, _) => {
+                                let block_expr = init.peel_blocks();
```

Comment:
I meant you call `peel_blocks` and just do `if let ExprKind::Tup(init_tup) = &init.peel_blocks().kind`. That way you don't have to worry about matching the block cass at all. With that you don't need to have duplicate arms or a closure either. Sorry for not being clear about this.